server/project.go

```diff
@@ -581,40 +587,44 @@ func (s *Server) composeProjectByID(ctx context.Context, id int) (*api.Project,
 	projectFind := &api.ProjectFind{
 		ID: &id,
 	}
-	project, err := s.ProjectService.FindProject(ctx, projectFind)
+	projectRaw, err := s.ProjectService.FindProject(ctx, projectFind)
 	if err != nil {
 		return nil, err
 	}
-	if project == nil {
+	if projectRaw == nil {
 		return nil, &common.Error{Code: common.NotFound, Err: fmt.Errorf("project ID not found: %d", id)}
 	}
 
-	if err := s.composeProjectRelationship(ctx, project); err != nil {
+	project, err := s.composeProjectRelationship(ctx, projectRaw)
+	if err != nil {
 		return nil, err
 	}
 
 	return project, nil
 }
 
-func (s *Server) composeProjectRelationship(ctx context.Context, project *api.Project) error {
+func (s *Server) composeProjectRelationship(ctx context.Context, projectRaw *api.ProjectRaw) (*api.Project, error) {
 	var err error
 
+	project := &api.Project{}
+	projectRaw.CopyToProject(project)
```

Comment:
CopyToProject is okay for now. However, we might want to find a more controlled pattern to limit the "copy" usage. The goal is to differentiate composed and raw object. So whatever we get for composed Project should be composed, and it's better to have only one place to do the composing process.

I'm thinking of having a composer library that takes an ID, e.g. getProject(id int), and returns the composed Project directly.

Reply:
> I'm thinking of having a composer library that takes an ID, e.g. getProject(id int), and returns the composed Project directly.

That's a good idea! I'll leave a note here.

Reply:
BTW, do you mean this
https://github.com/bytebase/bytebase/blob/main/server/project.go#L599
@d-bytebase

Reply:
Correct. All server code should use composed objects. I'm thinking of the following structure.

api package: only has composed object type, e.g. Project. We should not define ProjectRaw here.
store: define projectRaw here, but private. Add composer class (we need to find a better name) that reads raw objects and compose them to composed objects such as Project.
server: should always call composer class only and have no access to these private raw objects.

Reply:
@d-bytebase tracked in the tracking issue https://github.com/bytebase/bytebase/issues/758